packages/graph/hash_graph/lib/graph/src/store/postgres/resolve.rs

```diff
@@ -0,0 +1,161 @@
+        uri: &VersionedUri,
+    ) -> Result<Record<DataType>, QueryError>;
+
+    async fn read_all_property_types(&self) -> Result<RowStream, QueryError>;
+
+    async fn read_versioned_property_type(
+        &self,
+        uri: &VersionedUri,
+    ) -> Result<Record<PropertyType>, QueryError>;
+}
+
+#[async_trait]
+impl<T> PostgresContext for &T
+where
+    T: PostgresContext + Sync,
+{
+    async fn read_all_data_types(&self) -> Result<RowStream, QueryError> {
+        PostgresContext::read_all_data_types(*self).await
+    }
+
+    async fn read_versioned_data_type(
+        &self,
+        uri: &VersionedUri,
+    ) -> Result<Record<DataType>, QueryError> {
+        PostgresContext::read_versioned_data_type(*self, uri).await
+    }
+
+    async fn read_all_property_types(&self) -> Result<RowStream, QueryError> {
+        PostgresContext::read_all_property_types(*self).await
+    }
+
+    async fn read_versioned_property_type(
+        &self,
+        uri: &VersionedUri,
+    ) -> Result<Record<PropertyType>, QueryError> {
+        PostgresContext::read_versioned_property_type(*self, uri).await
+    }
+}
+
+#[derive(Debug)]
+pub struct Record<T> {
+    pub record: T,
+    pub is_latest: bool,
+}
+
+async fn read_all_types(client: &impl AsClient, table: &str) -> Result<RowStream, QueryError> {
+    client
+        .as_client()
+        .query_raw(
+            &format!(
+                r#"
+                SELECT schema, created_by, MAX(version) OVER (PARTITION by base_uri) = version as latest
+                FROM {table}
+                INNER JOIN ids
+                ON {table}.version_id = ids.version_id
+                ORDER BY base_uri, version DESC;
+                "#,
+            ),
+            parameter_list([]),
+        )
+        .await
+        .into_report()
+        .change_context(QueryError)
+}
+
+async fn read_versioned_type(
+    client: &impl AsClient,
+    table: &str,
+    uri: &VersionedUri,
+) -> Result<Record<Row>, QueryError> {
+    let type_row = client
+        .as_client()
+        .query_one(
+            &format!(
+                r#"
+                SELECT schema, created_by
+                FROM {table}
+                INNER JOIN ids
+                ON {table}.version_id = ids.version_id
+                WHERE base_uri = $1 AND version = $2;
+                "#
+            ),
+            &[&uri.base_uri().as_str(), &i64::from(uri.version())],
+        )
+        .await
+        .into_report()
+        .change_context(QueryError)?;
+
+    let row = client
+        .as_client()
+        .query_one(
+            r#"
+            SELECT MAX(version) as latest
+            FROM ids
+            WHERE base_uri = $1
+            "#,
+            &[&uri.base_uri().as_str()],
+        )
+        .await
+        .into_report()
+        .change_context(QueryError)?;
+    let latest: i64 = row.get(0);
+    Ok(Record {
+        record: type_row,
+        is_latest: latest as u32 == uri.version(),
+    })
```

Comment:
(untested) this would be what we'd do to have one query instead of two here.
We don't need to apply this change, but if you don't like having multiple queries, we can find a solution
```suggestion
    let type_row = client
        .as_client()
        .query_one(
            &format!(
                r#"
                SELECT schema, created_by, (
                    SELECT MAX(version) as latest 
                    FROM ids 
                    WHERE base_uri = $1
                    LIMIT 1
                )
                FROM {table} type_table
                INNER JOIN ids
                ON type_table.version_id = ids.version_id
                WHERE base_uri = $1 AND version = $2;
                "#
            ),
            &[&uri.base_uri().as_str(), &i64::from(uri.version())],
        )
        .await
        .into_report()
        .change_context(QueryError)?;

    let latest: i64 = type_row.get(2);
    Ok(Record {
        record: type_row,
        is_latest: latest as u32 == uri.version(),
    })
```

Reply:
Awesome, thanks!